---
title: "LUADMutationSummary"
author: "C.J. Liu"
date: "1/19/2017"
output:
  html_document:
    depth: 3
    highlight: tango
    number_sections: yes
    theme: united
    toc: yes
    toc_float:
      collapsed: no
      smooth_scroll: no
    word_document:
      toc: yes
  word_document:
    toc: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

```{r load library}
suppressMessages(require(dplyr))
suppressMessages(require(ggplot2))

```


## Somatic Mutation Calling

1. Download all cases of LUAD WXS bam files from GDC-portal.
2. Download Regulatory Build annotation from Ensembl.
3. Call Regulatory element mutations by GATK (DP > 10, AD > 3 )
4. Exclude dbSNP, avsnp147 (ANNOVAR)
5. Find recurrent tumor and nomal respectively.
6. Somatic tumor.recurrent - normal.recurrent.

## Load data
```{r load somatic}
dataPath <- '/home/cliu18/liucj/projects/1.Mutation_calling_in_non-condig_region_through_EXOME/4.compare_tumor_normal/LUAD/SNP'

header <-c("chrom","pos","ref","alt","ensid","desc","feature","recurrent")

chromOrder = c("chr1","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr20","chr21","chr22","chrX")

somatic <- as_data_frame(read.table(file.path(dataPath,'LUAD.somatic.recurrency'), header = F, sep = "\t"))

colnames(somatic) <- header

somatic <- filter(somatic, ensid != "." & recurrent > 1) %>% arrange(-recurrent)

somatic$chrom <- factor(somatic$chrom, levels = chromOrder)
```

### Recurrent count and feature distribution
```{r Recurrent and feature count}
ggplot(somatic, aes(x = factor(recurrent))) + geom_bar(aes(fill = feature)) + geom_text(stat = 'count', aes(label = ..count..), vjust = -0.3) + theme_minimal() + xlab("Recurrent") + ylab("Count") + ggtitle(paste("Total ",nrow(somatic)," (Recurrent > 2)"))
```

### Feature distribution on the chromosome
```{r}
ggplot(somatic, aes(x = chrom)) + geom_bar(aes(fill = feature)) + geom_text(stat = 'count', aes(label = ..count..), vjust = -0.3, size = 3) + theme_minimal() + theme(axis.text.x = element_text(angle = 45, hjust = 1)) + xlab("Chromosome") + ylab("Count") + ggtitle("Chromosome Distribution")
```

### Mutation type
```{r}
ggplot(somatic, aes(ref,alt)) + geom_bin2d() + theme_minimal() + scale_fill_gradient(low = "#00AFBB", high = "#FC3E07")
```

### Filter recurrent larger than or equal to 5
```{r}
somatic5 <- filter(somatic, recurrent >= 5)
#ggplot(somatic5, aes(x = ensid)) + geom_bar(aes(fill = feature)) + geom_text(stat = 'count', aes(label = ..count..), vjust = -0.3) + theme_minimal() + xlab("Recurrent") + ylab("Count") + ggtitle(paste("Total ",nrow(somatic5)," (Recurrent > 2)")) + theme(axis.text.x = element_text(angle = 90, hjust = 1))

#write the somatic 5 to file for testing 

write.table(somatic5, file = file.path(dataPath,"LUAD.somatic5.recurrency"),sep = "\t", quote = F, row.names = F, col.names = T)

ggplot(somatic5, aes(x = factor(recurrent))) + geom_bar(aes(fill = feature)) + geom_text(stat = 'count', aes(label = ..count..), vjust = -0.3) + theme_minimal() + xlab("Recurrent") + ylab("Count") + ggtitle(paste("Total ",nrow(somatic5)," (Recurrent >= 5)"))

ggplot(somatic5, aes(x = chrom)) + geom_bar(aes(fill = feature)) + geom_text(stat = 'count', aes(label = ..count..), vjust = -0.3, size = 3) + theme_minimal() + theme(axis.text.x = element_text(angle = 45, hjust = 1)) + xlab("Chromosome") + ylab("Count") + ggtitle("Chromosome Distribution")

ensCount <- somatic5 %>% group_by(ensid) %>% summarise(count = n()) %>% arrange(-count)
#kable(ensCount)
```

